src/acl/AnnotateClient.cc

```diff
@@ -17,15 +17,29 @@ int
 Acl::AnnotateClientCheck::match(ACLChecklist * const ch)
 {
     const auto checklist = Filled(ch);
+    auto annotated = false;
+    const auto tdata = dynamic_cast<ACLAnnotationData*>(data.get());
+    assert(tdata);
+    const auto conn = checklist->conn();
 
-    if (const auto conn = checklist->conn()) {
-        const auto tdata = dynamic_cast<ACLAnnotationData*>(data.get());
-        assert(tdata);
+    if (conn) {
         tdata->annotate(conn->notes(), &delimiters.value, checklist->al);
-        if (const auto request = checklist->request)
-            tdata->annotate(request->notes(), &delimiters.value, checklist->al);
-        return 1;
+        annotated = true;
     }
-    return 0;
+
+    if (const auto &request = checklist->request) {
+        tdata->annotate(request->notes(), &delimiters.value, checklist->al);
+        annotated = true;
+    } else if (conn && !conn->pipeline.empty()) {
+        debugs(28, DBG_IMPORTANT, "ERROR: Squid BUG: " << name << " ACL is used in context with " <<
+               "an unexpectedly nil ACLFilledChecklist::request. Did not annotate the current transaction.");
+    }
+
+    if (!annotated) {
+        debugs(28, DBG_IMPORTANT, "WARNING: " << name << " ACL is used in context without " <<
+               "active client-to-Squid connection and current transaction information. Did not annotate.");
```

Comment:
Acl::Node::matches() has similar "used in context without X" warnings, but there is an important difference that explains why this particular warning is here rather than in Acl::Node::matches(): Unlike Acl::Node::matches() warning cases, these problems do _not_ and must not lead to ACL mismatches. This warning is not about our inability to calculate AnnotateClientCheck::match() return value -- that value is always `1`. The warnings in Acl::Node::matches() say "Assuming mismatch". This warning says "Did not annotate" (and this ACL then "matches" by returning `1`).

This comment annotates this PR. It does not request any PR changes.